### DISCUSSION TITLE
compilation times

### DISCUSSION THREAD

**Discussion by thijssteel** [tlapack/tlapack#144] [Ideas]
A typical problem of c++ libraries is compilation time.

We're already running into this with the compilation time of the tester. Some solutions:

- Explicit template instantiation to get some precompilation
- Cool it a little with the templates, since more templates = extra compilation times (yes, i'm still against the side and op types ;) )
- Instead of including <tlapack.h> only include the necessary headers in the tests and examples.

**Comment by langou** [comment]
All good ideas @thijssteel.

"Explicit template instantiation to get some precompilation" => A little against. Main reason: this can wait. We might need to do this at some point though, I agree. "Cool it a little with the templates" => Yes, maybe. Same thing, maybe for now this can wait. But I see your point. This would help. Instead of compiling 2^4=16 versions of TRSM we would compile 1. This is mostly for our tests though, our tests test all 16 of them. I would assume a user code would only use a handful of these. Although, well, it goes fast. Indeed, someone easily uses quite a few of them. "Instead of including <tlapack.h> only include the necessary headers in the tests and examples" => Yes. We can try. Good idea for sure. This is the one idea I am the most in favor of.

Also to add to the list, one issue is ./tester.exe in itself, and it has all our unit tests, and we do not have many. This is not scalable. If we double the number of unit tests (which hopefully we will soon), we double the compile time. And, since this is a build that is done repeatedly during debug mode, development mode, it's really a pain for developer. We are small and it is already biting us. Can we break ./tester.exe in one executable per unit test?

**Comment by weslleyspereira** [comment]
> A typical problem of c++ libraries is compilation time.

Yes, thanks for raising the discussion!

>     * Explicit template instantiation to get some precompilation

I don't think this would help, but we can time it.

>     * Cool it a little with the templates, since more templates = extra compilation times (yes, i'm still against the side and op types ;) )

This could help a bit in terms of compilation time. But I think we should think more on usage than compilation time, when taking decisions on this thread. I have in mind:
- Every vector and matrix in the interface of a function can belong to a different class.
- Parameters like Op and Uplo could be templated with trans_t and uplo_t to allow the benefits of compile-time conditionals. We can improve this design, sure. Let's talk more about it! :)
- Integers, sizes and constants passed through the interface of a function don't need to have its own template type. We can certainly cool things here.

>     * Instead of including <tlapack.h> only include the necessary headers in the tests and examples.

This is the most relevant I think.

**Comment by weslleyspereira** [comment]
> Also to add to the list, one issue is ./tester.exe in itself, and it has all our unit tests, and we do not have many. This is not scalable. If we double the number of unit tests (which hopefully we will soon), we double the compile time. And, since this is a build that is done repeatedly during debug mode, development mode, it's really a pain for developer. We are small and it is already biting us. Can we break ./tester.exe in one executable per unit test?

We can have many executable, no problem with that. The advantage of having a single big tester is that we can do something like: `./tester.exe '[eigenvalues]'` which will run all tests tagged for eigenvalues. Solution: We provide separate testers, and add an option in the building system to enable/disable the compilation of the big `tester.exe`.

**Comment by thijssteel** [comment]
> Solution: We provide separate testers, and add an option in the building system to enable/disable the compilation of the big tester.exe.

I like this idea.

**Comment by thijssteel** [comment]
Actually, when looking at the compilation, most of the time is spent building the .o files for the tests. Linking them together into the tester isn't that much work. If we make sure that only the relevant files are included, the building should be fast.

**Comment by weslleyspereira** [comment]
> Actually, when looking at the compilation, most of the time is spent building the .o files for the tests. Linking them together into the tester isn't that much work. If we make sure that only the relevant files are included, the building should be fast.

I tried this solution here: https://github.com/tlapack/tlapack/pull/143, but the compilation times didn't change that much.

I also isolated the installation of Catch2. On Windows and MacOS, this may take close to 1min.

**Comment by weslleyspereira** [comment]
> Also to add to the list, one issue is ./tester.exe in itself, and it has all our unit tests, and we do not have many. This is not scalable. If we double the number of unit tests (which hopefully we will soon), we double the compile time. And, since this is a build that is done repeatedly during debug mode, development mode, it's really a pain for developer. We are small and it is already biting us. Can we break ./tester.exe in one executable per unit test?

I forgot to comment on that. Tests were broken at some point, I forgot the number of the PR. And, with #143, the tests start to depend only on the files they use. Good news for us, developers!

**Comment by weslleyspereira** [comment]
I converted this thread to a discussion so we do not lose the discussion about:

- Explicit template instantiation to get some precompilation
- Cool it a little with the templates, since more templates = extra compilation times

**Comment by weslleyspereira** [comment]
Related to https://github.com/tlapack/tlapack/pull/296.

**Comment by weslleyspereira** [comment]
> Explicit template instantiation to get some precompilation

This change does reduce the build time of the targets in 4%. See https://github.com/tlapack/tlapack/pull/296#issuecomment-1747302672

**Comment by weslleyspereira** [comment]
Another idea (from @thijssteel): Use caching in the CI, e.g., https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows
